tests/capi/unittest_capi_service_training_offloading.cc

```diff
@@ -0,0 +1,507 @@
+  double *output = NULL;
+  size_t data_size = 0;
+  int status;
+  double result_data[4];
+  char *output_node_name = NULL;
+
+  switch (event) {
+    case ML_SERVICE_EVENT_NEW_DATA:
+      ASSERT_TRUE (event_data != NULL);
+
+      status = ml_information_get (event_data, "name", (void **) &output_node_name);
+      EXPECT_EQ (status, ML_ERROR_NONE);
+      EXPECT_STREQ (output_node_name, "training_result");
+
+      status = ml_information_get (event_data, "data", &data);
+      EXPECT_EQ (status, ML_ERROR_NONE);
+
+      status = ml_tensors_data_get_tensor_data (data, 0, (void **) &output, &data_size);
+      EXPECT_EQ (status, ML_ERROR_NONE);
+      break;
+    default:
+      break;
+  }
+
+  if (output) {
+    for (int i = 0; i < 4; i++)
+      result_data[i] = output[i];
+
+    g_debug ("name:%s >> [training_loss: %f, training_accuracy: %f, validation_loss: %f, validation_accuracy: %f]",
+        output_node_name, result_data[0], result_data[1], result_data[2],
+        result_data[3]);
+  }
+}
+
+/**
+ * @brief Start thread
+*/
+static void
+sender_start_thread (ml_service_h sender_h)
+{
+  ml_service_start (sender_h);
+}
+
+/**
+ * @brief Client thread
+*/
+static void
+receiver_start_thread (ml_service_h receiver_h)
+{
+  ml_service_start (receiver_h);
+}
+
+/**
+ * @brief full test for training offloading services.
+ */
+TEST_F (MLServiceTrainingOffloading, trainingOffloading_p)
+{
+  ml_service_h receiver_h;
+  ml_service_h sender_h;
+
+  GThread *start_thread = NULL;
+  GThread *receive_thread = NULL;
+  const gchar *root_path = g_getenv ("MLAPI_SOURCE_ROOT_PATH");
+  g_autofree gchar *file_path
+      = g_build_filename (root_path, "tests", "test_models", "models", NULL);
+  ASSERT_TRUE (g_file_test (file_path, G_FILE_TEST_IS_DIR));
+  g_autofree gchar *trained_model_path = g_build_filename (
+      root_path, "tests", "test_models", "models", "trained-model.bin", NULL);
+
+  /* If you run the sender without running the receiver first, a connect error occurs in nns-edge. */
+  g_autofree gchar *receiver_config = _get_config_path ("training_offloading_receiver.conf");
+  status = ml_service_new (receiver_config, &receiver_h);
+  ASSERT_EQ (status, ML_ERROR_NONE);
+
+  g_autofree gchar *sender_config = _get_config_path ("training_offloading_sender.conf");
+  status = ml_service_new (sender_config, &sender_h);
+  ASSERT_EQ (status, ML_ERROR_NONE);
+
+  /* A path to app rw path */
+  status = ml_service_set_information (sender_h, "path", file_path);
+  /* A path to app rw path */
+  status = ml_service_set_information (receiver_h, "path", file_path);
+
+  status = ml_service_set_event_cb (sender_h, _receive_trained_model_cb, NULL);
+  EXPECT_EQ (status, ML_ERROR_NONE);
+
+  status = ml_service_set_event_cb (receiver_h, _sink_register_cb, NULL);
+  EXPECT_EQ (status, ML_ERROR_NONE);
+
+  start_thread = g_thread_new (
+      "sender_start_thread", (GThreadFunc) sender_start_thread, sender_h);
+  receive_thread = g_thread_new (
```

Comment:
Q: why do you use thread to start / stop training?

Reply:
If the test app uses the same thread as the receiver and sender,
If the receiver does lock for wait required file, the sender will not work.